### ISSUE TITLE
Train on ASPEC to get baseline performance

### ISSUE THREAD

**Issue by dennybritz** [rpryzant/japanese_corpus#9]
I preprocessed the ASPEC corpus to bring it into the right format:

- https://drive.google.com/open?id=0B_bZck-ksdkpYTkzYUI1cDJOanM

I used [google/sentencepiece](https://github.com/google/sentencepiece) to learn a shared wordpiece vocabulary.

Exact preprocessing:

```bash
# Split data into parallel files
cat ${ASPEC_PATH}/ASPEC-JE/train/*.txt | sed -e 's/\s|||\s/\t/g' | cut -f 4 > ${OUTPUT_DIR}/train.ja
cat ${ASPEC_PATH}/ASPEC-JE/train/*.txt | sed -e 's/\s|||\s/\t/g' | cut -f 5 > ${OUTPUT_DIR}/train.en

cat ${ASPEC_PATH}/ASPEC-JE/dev/*.txt | sed -e 's/\s|||\s/\t/g' | cut -f 3 > ${OUTPUT_DIR}/dev.ja
cat ${ASPEC_PATH}/ASPEC-JE/dev/*.txt | sed -e 's/\s|||\s/\t/g' | cut -f 4 > ${OUTPUT_DIR}/dev.en

cat ${ASPEC_PATH}/ASPEC-JE/test/*.txt | sed -e 's/\s|||\s/\t/g' | cut -f 3 > ${OUTPUT_DIR}/test.ja
cat ${ASPEC_PATH}/ASPEC-JE/test/*.txt | sed -e 's/\s|||\s/\t/g' | cut -f 4 > ${OUTPUT_DIR}/test.en

cat ${ASPEC_PATH}/ASPEC-JE/devtest/*.txt | sed -e 's/\s|||\s/\t/g' | cut -f 3 > ${OUTPUT_DIR}/devtest.ja
cat ${ASPEC_PATH}/ASPEC-JE/devtest/*.txt | sed -e 's/\s|||\s/\t/g' | cut -f 4 > ${OUTPUT_DIR}/devtest.en

# Install google/sentencepiece: https://github.com/google/sentencepiece

# Learn shared BPE
spm_train --input=${OUTPUT_DIR}/train.ja,${OUTPUT_DIR}/train.en -model_prefix=${OUTPUT_DIR}/bpe --vocab_size=32000 --model_type=bpe

# Apply shared BPE
for data in train dev test devtest; do
  spm_encode --model=${OUTPUT_DIR}/bpe.model --output_format=piece < ${OUTPUT_DIR}/${data}.en > ${OUTPUT_DIR}/${data}.bpe.en
  spm_encode --model=${OUTPUT_DIR}/bpe.model --output_format=piece < ${OUTPUT_DIR}/${data}.ja > ${OUTPUT_DIR}/${data}.bpe.ja
done
```

**Comment by dennybritz** [comment]
Will be training on this soon.

**Comment by rpryzant** [comment]
sentencepiece looks very cool.  I used [RakutenMA](https://github.com/rakuten-nlp/rakutenma) for JA tokenization on our data, but I'll retokenize it for a fairer comparison.

**Comment by dennybritz** [comment]
Yeah, tokenization can make a big difference. But BLEU scores on one corpus aren't that meaningful to compare to another corpus anyway. I'm training on ASPEC now, let's see what it converges to (hopefully tomorrow).

**Comment by rpryzant** [comment]
Why aren't they meaningful? Couldn't we make a common test set?

**Comment by dennybritz** [comment]
Well, BLEU on one dataset isn't comparable to BLEU on another dataset, just as something like classification accuracy depends on the dataset. If the data is noisy a model won't get a good BLEU score on it. We can combine them for testing, but then we can no longer compare to any published results because we don't know what the effect of the new data is. I think a betterstrategy may be to combine the training sets and see if we can use the new data to improve the one we have benchmarks for. Adding the new dat may actually *hurt* performance on the ASPEC eval because it's such a different domain...

**Comment by rpryzant** [comment]
Makes sense. 

I've updated https://github.com/rpryzant/japanese_corpus/issues/4 with the cleaned & processed 1.6M subset of our data.

**Comment by dennybritz** [comment]
I got some ASPEC predictions, but I cannot evaluate BLEU because the data is not tokenized (I used word pieces). I think you've done the tokenization before and read the papers that do it, could to try it on these predictions and see what BLEU you get? https://gist.github.com/dennybritz/7654310862b7522a9e665fa74d6fa849

**Comment by rpryzant** [comment]
See [[HERE]](https://gist.github.com/rpryzant/9fbcae5c321a827fa652caabd197e39c) for the tokenized files.

A respectable BLEU of 25.20.

**Comment by dennybritz** [comment]
Cool, thanks. By the way, I'm pretty sure that the papers with >30 BLEU do something different, evaluate on different data, or with different scripts. Would be worth looking into.

**Comment by rpryzant** [comment]
I think a lot of those papers did character-level chunking. That might explain some of it...you're bound to get more overlapping n-grams if each n-gram is defined over characters.

**Comment by dennybritz** [comment]
I followed the instruction from WAT above and got this:

```
BLEU = 36.11, 66.9/43.8/30.5/21.9 (BP=0.965, ratio=0.965, hyp_len=51946, ref_len=53809)
```

**Comment by rpryzant** [comment]
We could get a WAT paper out of that score alone :) 

How were their instructions different? What were your hyperparams?

**Comment by dennybritz** [comment]
These are the exact instructions: http://lotus.kuee.kyoto-u.ac.jp/WAT/evaluation/automatic_evaluation_systems/automaticEvaluationJA.html

I used the standard model from my NMT codebase: https://github.com/google/seq2seq/blob/master/example_configs/nmt_large.yml

It's probably largely due to using sentence pieces, I feel like that makes a large difference for Japanese.

**Comment by dennybritz** [comment]
The kytea model file I used is different from theirs (I couldn't find their and couldn't get that to work). That may make a bit of a difference but probably not a whole lot.

**Comment by rpryzant** [comment]
Agreed, sentence pieces is probably a big plus for JA-EN. Nice thinking there.

And yeah I played around with various kytea models when I was making my tokenizer. Didn't find them to be that different. kytea does both tokenization AND pos tagging, and I found that most models changed pos tagging behavior more than tokenization.

**Comment by rpryzant** [comment]
I think state-of-the-art is a plenty reasonable baseline haha...marking this as done.
